pep-9999.txt

```diff
@@ -0,0 +1,200 @@
+directory inside), then it will be used, otherwise the Python will behave as it
+does currently.
+
+
+If any package management tool finds the same ``__pypackages__`` directory in
+the current working directory, it will install any packages there and also
+create it if required based on Python version.
+
+
+Projects that use a source management system can include a ``__pypackages__``
+directory (empty or with e.g. a file like ``.gitignore``). After doing a fresh
+check out the source code, ``pip`` can be used to install the required
+dependencies directly into this directory.
+
+Example
+-------
+
+The following shows an example project directory structure, and different ways
+the Python executable and any script will behave.
+
+::
+
+    foo
+        __pypackages__
+            python3.8
+                Lib
+                    bottle
+        myscript.py
+
+    /> python foo/myscript.py
+    sys.path[0] == 'foo'
+    sys.path[x] == 'foo/__pypackages__/python3.8/Lib'
+    
+
+    cd foo
+
+    foo> /usr/bin/randomproject.sh
+        python /usr/share/randomproject/main.py
+
+    foo> /usr/bin/ansible  
+        #! /usr/bin/env python3
+    foo> python /usr/bin/ansible
+
+
+    foo> python myscript.py
+
+
+    foo> python
+    sys.path[0] == '.'
+    sys.path[x] == './__pypackages__/python3.8/Lib'
+
+    foo> python -m bottle
+
+
+    
+
+We have a project directory called ``foo`` and it has a ``__pypackages__``
+inside of it. We have ``bottle`` installed in that
+``__pypackages__/python3.8/Lib``, and have ``myscript.py`` file inside of the
+project directory.
+
+For invoking a script, for example ``/usr/bin/randomproject.sh``, Python will
+try to find a ``__pypackages__`` inside of the directory that the script
+resides[1]_, ``/usr/bin``.  The same will happen in case of the last example,
+where we are executing ``/usr/bin/ansible`` from inside of the ``foo``
+directory. In both cases, it will **not** use the ``__pypackages__`` in the
+current working directory.
+
+Similarly, if we invoke ``myscript.py`` from the first example, it will use the
+``__pypackages__ `` directory that was in the ``foo`` directory.
+
+
+If we go inside of the ``foo`` directory and start the Python executable (the
+interpreter), it will find the ``__pypackages__`` directory inside of the
+current working directory and use it in the ``sys.path``. The same happens if we
+try to use the ``-m`` and use a module. In our example, ``bottle`` module will
+be found inside of the ``__pypackages__`` directory.
+
+The above two examples are only cases where ``__pypackages__`` from current
+working directory is used.
+
+.. [1]_: In the case of symlinks, it is the directory where the actual script
+   resides, not the symlink pointing to the script
+
+
+Security Considerations
+=======================
+
+While executing a Python script, it will not consider the ``__pypackages__`` in
+the current directory, instead if there is a ``__pypackages__`` directory in the
```

Comment:
Can we maybe special-case out `/usr/bin` or similar directories somehow? It feels like 99% of the time this is going to result in a hundred extra `stat` calls per import (syscalls are slow again, thanks spectre & meltdown), with no benefit for any installed software.

Reply:
Do remember that importlib caches directory contents so stat calls are heavily curtailed.